dependent_types.md

```diff
@@ -0,0 +1,386 @@
+
+## Terms vs. Types
+Difference between the terms and types can be looked at in a few different ways. 
+
+First one would be that if you look at a type/kind annotation:
+```
+term :: type :: kind
+```
+_term_ is the thing to the left of the type annotation (`::`) and _type_ is to the right of it.
+You can ignore kinds for now, they are to types what types are to terms - so like type of a type constructor. You can think of them as _types one level up_. Read about kinds [here](https://wiki.haskell.org/Kind)
+
+We can also say something that applies to standard Haskell (without extensions) and that is that terms are present at runtime while types get erased at runtime. This statement is not true when it comes to dependent types but it serves well in the path of our understanding of the topic.
+
+## Type level data
+We can say for a type that it has a set of possible values that correspond to it. So `Void` is a type with zero inhabitants - empty set , `Unit` has a single element set (`()`) , `Bool` has two element set (`True` and `False`) and so on. Here is a small example of this
+
+```
+data Void 
+data Unit = ()
+data Bool = True | False
+```
+
+In contrast to this types can also contain _type level data_. That is the data that lives on the type level and does not have associated set of inhabitants. 
+
+Here is a small example that provides a way of defining a type level _String_ (`Symbol`)
+
+```
+data Label (l :: Symbol) = Get
+```
+
+Here type `Label` accepts a type parameter `l` which must be of type `Symbol`. Type level data is something that lives on the level of types but is not a type.
+
+## Lambda Cube
+
+[Lambda Cube](https://en.wikipedia.org/wiki/Lambda_cube) is the framework for exploring the forms of abstraction. Starting in some order from simple lambda calculus to calculus of constructions (higher-order dependently typed polymorphic lambda calculus) we first arrive to what every programming language has :
+
+**Values depending on values** (_functions_)
+- This simply means that you have some sort of relation between variables if they are used in terms of each other. 
+
+Example:
+
+```
+a :: Integer
+b = even a :: Bool
+-- here b depends on a 
+```
+
+After that we get to the second step which is
+
+**Values depending on Types** (_classes_)
+
+Example:
+
+```
+maxBound :: Int
+maxBound :: Double
+```
+
+We can see here that the type is determining the values `maxBound` will have. Values depend on the type they are instantiated with so the type to the right of the `::` determines the values to the left of the `::`.
+
+
+**Types depending on Types** (_type functions_)
+
+In order to demonstrate what we mean by this we will use Haskell extension called Type Families. 
+
+Type Families provide a way to define multiple type constructors with different types in place of a type parameter.
+Here is an example from Haskell [wiki](https://wiki.haskell.org/Haskell)
+
+```
+-- Declare a list-like data family
+data family XList a
```

Comment:
The section is about type families, why use a data family as an example?

Reply:
I think it's valid because a data family is an injective generative type family with `*` as codomain, but might be a bit confusing to the reader.